source/request-response/request_response_client.c

```diff
@@ -0,0 +1,388 @@
+    const struct aws_rr_subscription_status_event *event,
+    void *userdata) {
+    (void)event;
+    (void)userdata;
+
+    /*
+     * We must be on the event loop, but it's safer overall to process this event as a top-level event loop task.  The
+     * subscription manager assumes that we won't call APIs on it while iterating subscription records and listeners.
+     *
+     * These tasks hold an internal reference while they exist.
+     */
+
+    /* NYI */
+}
+
+static void s_aws_rr_client_protocol_adapter_subscription_event_callback(
+    const struct aws_protocol_adapter_subscription_event *event,
+    void *user_data) {
+    struct aws_mqtt_request_response_client *rr_client = user_data;
+
+    AWS_FATAL_ASSERT(aws_event_loop_thread_is_callers_thread(rr_client->loop));
+
+    if (rr_client->state != AWS_RRCS_ACTIVE) {
+        return;
+    }
+
+    aws_rr_subscription_manager_on_protocol_adapter_subscription_event(&rr_client->subscription_manager, event);
+}
+
+static void s_aws_rr_client_protocol_adapter_incoming_publish_callback(
+    const struct aws_protocol_adapter_incoming_publish_event *publish,
+    void *user_data) {
+    (void)publish;
+
+    struct aws_mqtt_request_response_client *rr_client = user_data;
+
+    AWS_FATAL_ASSERT(aws_event_loop_thread_is_callers_thread(rr_client->loop));
+
+    if (rr_client->state != AWS_RRCS_ACTIVE) {
+        return;
+    }
+
+    /* NYI */
+}
+
+static void s_aws_rr_client_protocol_adapter_terminate_callback(void *user_data) {
+    struct aws_mqtt_request_response_client *rr_client = user_data;
+
+    /* release the internal ref count "held" by the protocol adapter's existence */
+    aws_ref_count_release(&rr_client->internal_ref_count);
+}
+
+static void s_aws_rr_client_protocol_adapter_connection_event_callback(
+    const struct aws_protocol_adapter_connection_event *event,
+    void *user_data) {
+    struct aws_mqtt_request_response_client *rr_client = user_data;
+
+    AWS_FATAL_ASSERT(aws_event_loop_thread_is_callers_thread(rr_client->loop));
+
+    if (rr_client->state != AWS_RRCS_ACTIVE) {
+        return;
+    }
+
+    aws_rr_subscription_manager_on_protocol_adapter_connection_event(&rr_client->subscription_manager, event);
+}
+
+static struct aws_mqtt_request_response_client *s_aws_mqtt_request_response_client_new(
+    struct aws_allocator *allocator,
+    const struct aws_mqtt_request_response_client_options *options,
+    struct aws_event_loop *loop) {
+    struct aws_rr_subscription_manager_options sm_options = {
+        .max_subscriptions = options->max_subscriptions,
+        .operation_timeout_seconds = options->operation_timeout_seconds,
+    };
+
+    /*
+     * We can't initialize the subscription manager until we're running on the event loop, so make sure that
+     * initialize can't fail by checking its options for validity now.
+     */
+    if (!aws_rr_subscription_manager_are_options_valid(&sm_options)) {
```

Comment:
Note: Maybe it's already in place somewhere or we simply haven't gotten around to it yet but just noting that an `operation_timeout_seconds` of 0 results in this failing. Wondering if there's a point before this where a default is applied to that value if it hasn't been set or if that default should be applied during the validity check.

Reply:
Timeout support is next.  Based on our discussion, I'm thinking about the following:

* Split operation_timeout_seconds into (request_response_timeout_seconds and streaming_subscribe_timeout_seconds)
* A zero-valued request response timeout gets turned into a default (30 or 60)
* A zero-value streaming timeout means no timeout

Does the inconsistency between how zero is treated by each setting outweigh the fact that we really want to have a reasonable default for request-response timeout and no timeout default for streaming operations?